presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -179,14 +193,33 @@ private RelationPlan createTableWriterPlan(Analysis analysis, RelationPlan plan,
             source = new LimitNode(idAllocator.getNextId(), source, 0L);
         }
 
+        // todo this should be checked in analysis
+        writeTableLayout.ifPresent(layout -> {
+            if (!ImmutableSet.copyOf(columnNames).containsAll(layout.getPartitionColumns())) {
+                throw new PrestoException(NOT_SUPPORTED, "INSERT must write all distribution columns");
+            }
+        });
+
+        List<Symbol> symbols = plan.getOutputSymbols();
+        Optional<PartitionFunctionBinding> partitionFunctionBinding = writeTableLayout.map(layout -> new PartitionFunctionBinding(
+                layout.getPartitioning(),
+                symbols,
+                layout.getPartitionColumns().stream()
+                        .mapToInt(columnNames::indexOf)
+                        .mapToObj(symbols::get)
+                        .collect(toImmutableList()),
+                Optional.empty()));
+
         PlanNode writerNode = new TableWriterNode(
                 idAllocator.getNextId(),
                 source,
                 target,
-                plan.getOutputSymbols(),
+                symbols,
                 columnNames,
                 writerOutputs,
-                plan.getSampleWeight());
+                plan.getSampleWeight(),
+                writeTableLayout.map(CreateTableLayout::getPartitioning),
```

Comment:
Keeping the partitioningHandle separately seems unnecessary since the partitionFunctionBinding contains it, too.